internal/kibana/role.go

```diff
@@ -0,0 +1,506 @@
+										return diags
+									}
+								}
+								diag := diag.Diagnostic{
+									Severity: diag.Error,
+									Summary:  "Wrong value for base attribute",
+									Detail:   fmt.Sprintf("Expected %s , got %s", strings.Join(expected, " | "), value),
+								}
+								diags = append(diags, diag)
+								return diags
+							},
+						},
+					},
+					"feature": {
+						Description: "List of privileges for specific features. When the feature privileges are specified, you are unable to use the \"base\" section.",
+						Type:        schema.TypeSet,
+						Optional:    true,
+						Elem: &schema.Resource{
+							Schema: map[string]*schema.Schema{
+								"name": {
+									Description: "Feature name.",
+									Type:        schema.TypeString,
+									Required:    true,
+								},
+								"privileges": {
+									Description: "Feature privileges.",
+									Type:        schema.TypeSet,
+									Required:    true,
+									Elem: &schema.Schema{
+										Type: schema.TypeString,
+									},
+								},
+							},
+						},
+					},
+					"spaces": {
+						Description: "The spaces to apply the privileges to. To grant access to all spaces, set to [\"*\"], or omit the value.",
+						Type:        schema.TypeSet,
+						Required:    true,
+						Elem: &schema.Schema{
+							Type: schema.TypeString,
+						},
+					},
+				},
+			},
+		},
+		"metadata": {
+			Description:      "Optional meta-data.",
+			Type:             schema.TypeString,
+			Optional:         true,
+			Computed:         true,
+			ValidateFunc:     validation.StringIsJSON,
+			DiffSuppressFunc: utils.DiffJsonSuppress,
+		},
+	}
+
+	return &schema.Resource{
+		Description: "Creates a Kibana role. See, https://www.elastic.co/guide/en/kibana/master/role-management-api-put.html",
+
+		CreateContext: resourceRoleUpsert,
+		UpdateContext: resourceRoleUpsert,
+		ReadContext:   resourceRoleRead,
+		DeleteContext: resourceRoleDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: apikeySchema,
+	}
+}
+
+func resourceRoleUpsert(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	kibana, err := client.GetKibanaClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	queryParams := ""
+	if d.IsNewResource() {
+		queryParams = "?createOnly=true"
+	}
+	kibanaRole := kbapi.KibanaRole{
+		Name:          fmt.Sprintf("%s%s", d.Get("name").(string), queryParams),
+		Kibana:        []kbapi.KibanaRoleKibana{},
+		Elasticsearch: &kbapi.KibanaRoleElasticsearch{},
+	}
+
+	if v, ok := d.GetOk("kibana"); ok {
+		definedKibanaConfigs := v.(*schema.Set)
+		kibanaRole.Kibana = make([]kbapi.KibanaRoleKibana, definedKibanaConfigs.Len())
+		for i, item := range definedKibanaConfigs.List() {
+			each := item.(map[string]interface{})
+			_config := kbapi.KibanaRoleKibana{
```

Comment:
```suggestion
			config := kbapi.KibanaRoleKibana{ 
```

Can we remove the leading underscores